src/systems/filecoin_markets/storage_market/storage_market_actor_code.go

```diff
@@ -250,170 +208,166 @@ func (a *StorageMarketActorCode_I) TerminateDealsOnSlashProviderSector(rt Runtim
 	}
 
 	UpdateRelease(rt, h, st)
 }
 
-func (st *StorageMarketActorState_I) _rtUpdatePendingDealStatesForParty(rt Runtime, addr addr.Address) {
-	// For consistency with OnEpochTickEnd, only process updates up to the end of the _previous_ epoch.
-	epoch := rt.CurrEpoch() - 1
+func (a *StorageMarketActorCode_I) OnEpochTickEnd(rt Runtime) {
+	h, st := a.State(rt)
 
-	var relevantDealIDs []deal.DealID
-	// Populate with the set of all elements in st.Deals() in which addr is one of the parties
-	// (either client or provider).
+	// Some deals may never be affected by the normal calls to _rtUpdatePendingDealStatesForParty
+	// (notably, if the relevant party never checks its balance).
+	// Without some cleanup mechanism, these deals may gradually accumulate and cause
+	// the StorageMarketActor state to grow without bound.
+	// To prevent this, we amortize the cost of this cleanup by processing a relatively
+	// small number of deals every epoch, independent of the calls above.
+	var cleanupDealIDs []deal.DealID
+	// Populate with the N oldest deals (e.g., by a priority queue on EndEpoch).
+	// N is a system parameter TBD, which may be a function of the global statistics
+	// (including the number of deals in each prior epoch).
 	IMPL_FINISH()
 
-	st._rtUpdatePendingDealStates(rt, relevantDealIDs, epoch)
-}
+	amountSlashedTotal := st._updatePendingDealStates(cleanupDealIDs, rt.CurrEpoch())
 
-func (st *StorageMarketActorState_I) _rtUpdatePendingDealStates(rt Runtime, dealIDs []deal.DealID, epoch block.ChainEpoch) {
-	for _, dealID := range dealIDs {
-		st._rtUpdatePendingDealState(rt, dealID, epoch)
-	}
+	UpdateRelease(rt, h, st)
+
+	rt.SendFunds(addr.BurntFundsActorAddr, amountSlashedTotal)
 }
 
-func (st *StorageMarketActorState_I) _rtUpdatePendingDealState(rt Runtime, dealID deal.DealID, epoch block.ChainEpoch) {
-	deal, dealP := st._getOnChainDealAssert(dealID)
+////////////////////////////////////////////////////////////////////////////////
+// Method utility functions
+////////////////////////////////////////////////////////////////////////////////
 
-	everUpdated := (deal.LastUpdatedEpoch() != block.ChainEpoch_None)
-	everSlashed := (deal.SlashEpoch() != block.ChainEpoch_None)
+func (st *StorageMarketActorState_I) _rtUpdatePendingDealStatesForParty(rt Runtime, addr addr.Address) (
+	amountSlashedTotal actor.TokenAmount) {
 
-	Assert(!everUpdated || (deal.LastUpdatedEpoch() <= epoch))
-	if deal.LastUpdatedEpoch() == epoch {
-		return
-	}
+	// For consistency with OnEpochTickEnd, only process updates up to the end of the _previous_ epoch.
+	epoch := rt.CurrEpoch() - 1
 
-	if deal.SectorStartEpoch() == block.ChainEpoch_None {
-		// Not yet appeared in proven sector; check for timeout.
-		if dealP.StartEpoch() >= epoch {
-			st._rtProcessDealInitTimedOut(rt, dealID)
-		}
-		return
-	}
+	var relevantDealIDs []deal.DealID
+	// Populate with the set of all elements in st.Deals() in which addr is one of the parties
```

Comment:
We might not have an efficient way to do this, deals are not indexed by the parties. We can consider allowing this function to take in an input array of DealIDs.

Reply:
They're not indexed by the parties on-chain, but implementations can maintain a local index for this -- I'll add a note to this effect. (Though, we might want to consider partitioning the on-chain data structure by party also.)